// ngx-webrtc/esm2020/lib/directives/share-screen.directive.mjs
import { ChangeDetectorRef, Directive, HostBinding, HostListener } from '@angular/core';
import { fromEvent, merge } from 'rxjs';
import { take } from 'rxjs/operators';
import { CallService } from '../services/call.service';
import { StreamService } from '../services/stream.service';
import * as i0 from "@angular/core";
import * as i1 from "../services/stream.service";
import * as i2 from "../services/call.service";
/**
 * Trigger get capture screen permissions and send screen to CallService.
 * You can listen to the change and call replaceTrack of peer connection to send the screen capture to that connection
 */
export class ShareScreenDirective {
    constructor(streamService, callService, cdr) {
        this.streamService = streamService;
        this.callService = callService;
        this.cdr = cdr;
        this.desktopStream = null;
        this.isDisabled = true;
        this.isEnabled = false;
    }
    async onClick() {
        if (!this.isEnabled) {
            await this.startShareScreen();
        }
        else {
            this.stopShareScreen();
        }
    }
    async startShareScreen() {
        this.desktopStream = await this.streamService.getScreenCapture();
        if (this.desktopStream) {
            this.streamService.localShareScreenStream$.next(this.desktopStream);
            const desktopVideoTrack = this.streamService.getVideoTrackForStream(this.desktopStream);
            if (desktopVideoTrack) {
                this.streamService.replaceTrack(desktopVideoTrack);
            }
            else {
                console.warn('no videotrack in desktop stream');
            }
            const streamInactive$ = fromEvent(this.desktopStream, 'inactive').pipe(take(1));
            const sharingStopped$ = fromEvent(this.desktopStream.getVideoTracks()[0], 'ended').pipe(take(1));
            merge(streamInactive$, sharingStopped$)
                .pipe(take(1))
                .subscribe(() => {
                if (this.isEnabled) {
                    this.stopShareScreen();
                }
            });
            this.isEnabled = !this.isEnabled;
            this.isDisabled = !this.isDisabled;
            this.callService.startShareScreen.emit();
            this.cdr.detectChanges();
        }
    }
    stopShareScreen() {
        this.streamService.localShareScreenStream$.next(null);
        const videoTrack = this.streamService.getVideoTrackForStream();
        if (videoTrack) {
            this.streamService.replaceTrack(videoTrack);
        }
        else {
            console.warn('video track not found');
        }
        if (this.desktopStream) {
            this.desktopStream.getTracks().forEach(track => {
                track.stop();
            });
        }
        this.callService.stopShareScreen.emit();
        this.isEnabled = !this.isEnabled;
        this.isDisabled = !this.isDisabled;
        this.cdr.detectChanges();
    }
}
ShareScreenDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: ShareScreenDirective, deps: [{ token: i1.StreamService }, { token: i2.CallService }, { token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Directive });
ShareScreenDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.2.1", type: ShareScreenDirective, selector: "[ngxWebrtcShareScreen]", host: { listeners: { "click": "onClick($event)" }, properties: { "class.disabled": "this.isDisabled", "class.enabled": "this.isEnabled" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: ShareScreenDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngxWebrtcShareScreen]',
                }]
        }], ctorParameters: function () { return [{ type: i1.StreamService }, { type: i2.CallService }, { type: i0.ChangeDetectorRef }]; }, propDecorators: { isDisabled: [{
                type: HostBinding,
                args: ['class.disabled']
            }], isEnabled: [{
                type: HostBinding,
                args: ['class.enabled']
            }], onClick: [{
                type: HostListener,
                args: ['click', ['$event']]
            }] } });
