// ngx-webrtc/esm2020/lib/directives/toggle-audio-user.directive.mjs
import { Directive, HostBinding, HostListener, Input } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Send a toggle audio request to a specific peer connection.
 */
export class ToggleAudioUserDirective {
    constructor() {
        this.ngxWebrtcToggleAudioUser = null;
        this.isDisabled = true;
        this.isEnabled = false;
    }
    onClick() {
        this.toggleUserAudio();
    }
    toggleUserAudio() {
        if (!this.ngxWebrtcToggleAudioUser) {
            console.warn('user not set');
            return;
        }
        if (this.ngxWebrtcToggleAudioUser?.connection) {
            this.ngxWebrtcToggleAudioUser.connection.requestMuteAudio();
            this.isEnabled = !this.isEnabled;
            this.isDisabled = !this.isDisabled;
        }
    }
}
ToggleAudioUserDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: ToggleAudioUserDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
ToggleAudioUserDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.2.1", type: ToggleAudioUserDirective, selector: "[ngxWebrtcToggleAudioUser]", inputs: { ngxWebrtcToggleAudioUser: "ngxWebrtcToggleAudioUser" }, host: { listeners: { "click": "onClick($event)" }, properties: { "class.disabled": "this.isDisabled", "class.enabled": "this.isEnabled" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: ToggleAudioUserDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngxWebrtcToggleAudioUser]',
                }]
        }], propDecorators: { ngxWebrtcToggleAudioUser: [{
                type: Input
            }], isDisabled: [{
                type: HostBinding,
                args: ['class.disabled']
            }], isEnabled: [{
                type: HostBinding,
                args: ['class.enabled']
            }], onClick: [{
                type: HostListener,
                args: ['click', ['$event']]
            }] } });
