// ngx-webrtc/esm2020/lib/directives/toggle-audio-self.directive.mjs
import { ChangeDetectorRef, Directive, EventEmitter, HostBinding, HostListener, Output } from '@angular/core';
import { StreamType } from "../enums/stream-type";
import { CallService } from '../services/call.service';
import { StreamService } from '../services/stream.service';
import * as i0 from "@angular/core";
import * as i1 from "../services/stream.service";
import * as i2 from "../services/call.service";
/**
 * Toggle disabled/enable audio track to mute/unmute local audio.
 */
export class ToggleAudioSelfDirective {
    constructor(streamService, callService, cdr) {
        this.streamService = streamService;
        this.callService = callService;
        this.cdr = cdr;
        this.callStartedSubscription$ = null;
        this.localStreamStatusChangedSubscription$ = null;
        this.toggleMuteFailed = new EventEmitter();
        this.isDisabled = true;
        this.isEnabled = false;
        this.init();
    }
    onClick() {
        this.toggleMute();
    }
    ngOnDestroy() {
        this.callStartedSubscription$?.unsubscribe();
        this.localStreamStatusChangedSubscription$?.unsubscribe();
    }
    init() {
        this.callStartedSubscription$ = this.callService.started$.subscribe(this.onStart.bind(this));
    }
    onStart(isStarted) {
        if (isStarted) {
            this.localStreamStatusChangedSubscription$ = this.streamService.localStreamStatusChanged.subscribe(this.onLocalStreamStatusChanged.bind(this));
        }
    }
    onLocalStreamStatusChanged(stream) {
        if (stream) {
            if (stream instanceof MediaStreamTrack && stream.kind === StreamType.Audio) {
                this.updateStatusWithTrack(stream);
            }
            if (stream instanceof MediaStream) {
                const track = this.streamService.getAudioTrackForStream(stream);
                this.updateStatusWithTrack(track);
            }
        }
    }
    toggleMute() {
        this.streamService.toggleLocalTrack(StreamType.Audio).catch(error => {
            this.toggleMuteFailed.emit(error);
        });
    }
    updateStatusWithTrack(track) {
        this.isEnabled = track && track.enabled ? true : false;
        this.isDisabled = !track || !track.enabled;
        this.cdr.detectChanges();
    }
}
ToggleAudioSelfDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: ToggleAudioSelfDirective, deps: [{ token: i1.StreamService }, { token: i2.CallService }, { token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Directive });
ToggleAudioSelfDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.2.1", type: ToggleAudioSelfDirective, selector: "[ngxWebrtcToggleAudioSelf]", outputs: { toggleMuteFailed: "toggleMuteFailed" }, host: { listeners: { "click": "onClick($event)" }, properties: { "class.disabled": "this.isDisabled", "class.enabled": "this.isEnabled" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: ToggleAudioSelfDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngxWebrtcToggleAudioSelf]',
                }]
        }], ctorParameters: function () { return [{ type: i1.StreamService }, { type: i2.CallService }, { type: i0.ChangeDetectorRef }]; }, propDecorators: { toggleMuteFailed: [{
                type: Output
            }], isDisabled: [{
                type: HostBinding,
                args: ['class.disabled']
            }], isEnabled: [{
                type: HostBinding,
                args: ['class.enabled']
            }], onClick: [{
                type: HostListener,
                args: ['click', ['$event']]
            }] } });
