// ngx-webrtc/esm2020/lib/enums/peer-connection-client-signal-message-type.mjs
export var PeerConnectionClientSignalMessageType;
(function (PeerConnectionClientSignalMessageType) {
    PeerConnectionClientSignalMessageType["Candidate"] = "candidate";
    PeerConnectionClientSignalMessageType["Answer"] = "answer";
    PeerConnectionClientSignalMessageType["Offer"] = "offer";
    PeerConnectionClientSignalMessageType["Bye"] = "bye";
    PeerConnectionClientSignalMessageType["RequestMuteAudio"] = "request-mute-audio";
    PeerConnectionClientSignalMessageType["RequestMuteVideo"] = "request-mute-video";
    PeerConnectionClientSignalMessageType["AudioMuted"] = "audio-muted";
    PeerConnectionClientSignalMessageType["AudioUnmuted"] = "audio-unmuted";
    PeerConnectionClientSignalMessageType["VideoMuted"] = "video-muted";
    PeerConnectionClientSignalMessageType["VideoUnmuted"] = "video-unmuted";
    PeerConnectionClientSignalMessageType["StartShareScreen"] = "start-share-screen";
    PeerConnectionClientSignalMessageType["StopShareScreen"] = "stop-share-screen";
})(PeerConnectionClientSignalMessageType || (PeerConnectionClientSignalMessageType = {}));
