// ngx-webrtc/esm2020/lib/ngx-webrtc.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ShareScreenDirective, ToggleAudioSelfDirective, ToggleAudioUserDirective, ToggleVideoSelfDirective, ToggleVideoUserDirective } from './directives';
import { Configuration } from './ngx-webrtc-configuration';
import { NGX_WEBRTC_STORAGE } from './ngx-webrtc-storage';
import { NGX_WEBRTC_STORAGE_PREFIX } from './ngx-webrtc-storage-prefix';
import * as i0 from "@angular/core";
export class NgxWebrtcModule {
    static forRoot(libConfiguration) {
        return {
            ngModule: NgxWebrtcModule,
            providers: [
                {
                    provide: Configuration,
                    useValue: libConfiguration,
                },
            ],
        };
    }
}
NgxWebrtcModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: NgxWebrtcModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxWebrtcModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: NgxWebrtcModule, declarations: [ToggleAudioSelfDirective,
        ToggleVideoSelfDirective,
        ToggleAudioUserDirective,
        ToggleVideoUserDirective,
        ShareScreenDirective], imports: [CommonModule], exports: [ToggleAudioSelfDirective,
        ToggleVideoSelfDirective,
        ToggleAudioUserDirective,
        ToggleVideoUserDirective,
        ShareScreenDirective] });
NgxWebrtcModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: NgxWebrtcModule, providers: [
        { provide: NGX_WEBRTC_STORAGE_PREFIX, useValue: 'ngx-webrtc' },
        { provide: NGX_WEBRTC_STORAGE, useValue: 'sessionStorage' },
    ], imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: NgxWebrtcModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        ToggleAudioSelfDirective,
                        ToggleVideoSelfDirective,
                        ToggleAudioUserDirective,
                        ToggleVideoUserDirective,
                        ShareScreenDirective,
                    ],
                    exports: [
                        ToggleAudioSelfDirective,
                        ToggleVideoSelfDirective,
                        ToggleAudioUserDirective,
                        ToggleVideoUserDirective,
                        ShareScreenDirective,
                    ],
                    imports: [CommonModule],
                    providers: [
                        { provide: NGX_WEBRTC_STORAGE_PREFIX, useValue: 'ngx-webrtc' },
                        { provide: NGX_WEBRTC_STORAGE, useValue: 'sessionStorage' },
                    ]
                }]
        }] });
